**회귀 분석**
- 사례) 최종 성적(Y)을 출석률(X1)과 과제점수(X2)로 설명



In [2]:
# 회귀분석: 독립변수 x가 종속변수 y에 미치는 영향을 분석
# 결정계수 R의 2승: 회귀분석의 설명력

import pandas as pd

df = pd.DataFrame({
    "출석률": [90, 85, 88, 75, 80, 95, 70, 65, 92, 78],
    "과제점수": [85, 80, 82, 70, 75, 90, 68, 60, 88, 72],
    "성적": [88, 82, 85, 74, 78, 92, 70, 65, 90, 76]
})

1. 다중회귀분석 수행

In [ ]:
import statsmodels.formula.api as smf

# 종속변수: 성적
# 독립변수: 출석률, 과제점수
model = smf.ols('성적 ~ 출석률 + 과제점수', data=df).fit()

print(model.summary())

# 결정계수 R-squared: 전체 성적 변동의 99.8%를 출석률/과제점수로 설명 가능
# 조정된 결정계수 Adj. R-squared: 설명변수 개수를 보정한 값, 99.7% 설명

# 회귀계수 coef
# P 값... 0.007, 0.046 둘 다 유의미한 변수(유의수준 0.05보다 작음)
# 신뢰구간(p 뒤에 구간 차이 구하기)

# 회귀식: 성적 = 7.061(절편) + 0.551*출석률 + 0.362*과제점수


                            OLS Regression Results                            
Dep. Variable:                     성적   R-squared:                       0.998
Model:                            OLS   Adj. R-squared:                  0.997
Method:                 Least Squares   F-statistic:                     1502.
Date:                Tue, 09 Jun 2026   Prob (F-statistic):           6.06e-10
Time:                        18:07:50   Log-Likelihood:                -5.2390
No. Observations:                  10   AIC:                             16.48
Df Residuals:                       7   BIC:                             17.39
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.0613      1.396      5.058      0.0

2. 결과해석

(1) 결정계수(R-squared, Adj. R-squared)
- R-squared  = 0.998
  → 전체 성적 변동의 99.8%를 출석률과 과제점수 두 변수로 설명 가능
- Adj. R-squared = 0.997
  → 설명변수 개수를 보정한 값으로도 99.7% 설명 가능

(2) 회귀계수 (coef)
- Intercept = 7.0613
- 출석률 = 0.5512 (p=0.007)
  → 출석률이 1% 증가할 때 성적은 약 0.55점 증가
  → p < 0.05 이므로 통계적으로 유의미한 변수
- 과제점수 = 0.3617 (p=0.046)
  → 과제점수가 1점 증가할 때 성적은 약 0.36점 증가
  → p < 0.05 이므로 통계적으로 유의미한 변수

(3) 결론
- 회귀식 : 성적 = 7.061 + 0.551 x 출석률 + 0.362 x 과제점수


3. 상관계수 구하기

In [ ]:
df[["출석률", "과제점수"]].corr()
# corr()은 기본적으로 피어슨 상관계수

,출석률,과제점수
출석률,1.000000,0.993559
과제점수,0.993559,1.000000


In [5]:
# method 옵션으로 스피어만, 켄달 상관 계수 계산도 가능
# 다중공선성 있으면 안됨... 독립변수 간의 상관 관계

print('스피어만 상관 계수 : \n', df[["출석률", "과제점수"]].corr(method='spearman'))
print('켄달 상관 계수 : \n', df[["출석률", "과제점수"]].corr(method='kendall'))

스피어만 상관 계수 : 
       출석률  과제점수
출석률   1.0   1.0
과제점수  1.0   1.0
켄달 상관 계수 : 
       출석률  과제점수
출석률   1.0   1.0
과제점수  1.0   1.0


4. 출석률 82점, 과제점수 80점에 대해, 구축한 회귀 모델에서의 예상 성적 구하기

In [6]:
new_data = pd.DataFrame({"출석률": [82], "과제점수": [80]}) # 새 데이터를 데이터프레임으로 만든 후에 모델에 넣기
pred = model.predict(new_data)
print("예상 성적:", pred[0])

예상 성적: 81.19539061176069
